server/cpho/views/indicators.py

```diff
@@ -78,6 +81,16 @@ class ViewIndicator(TemplateView):
     model = Indicator
     template_name = "indicators/view_indicator.jinja2"
 
+    def get(self, request, *args, **kwargs):
+        context = self.get_context_data(**kwargs)
+        if not test_rule(
+            "can_edit_indicator", request.user, context["object"]
+        ):
+            raise SuspiciousOperation(
+                f"User {request.user} cannot edit this indicator"
+            )
+        return self.render_to_response(context)
+
```

Comment:
I think we'll instead want 1 or 2 "base" mixins and potentially even more if you find yourself repeating yourself often,

The more standard way of doing auth checks in CBVs is with the dispatch method, because it's defined on the base view class and is fired immediately before any other methods when a request is sent. It gets run regardless of GET/POST/PUT/etc.  

```python
class MustPassAuthCheckMixin(View):
    def check_rule(self):
        raise NotImplementedError("must override check_rule") 

    def dispatch(...)
          if not self.check_rule():
               raise PermissionDenied()

          return super().dispatch(..)          
          
```
Many rules don't take arguments, so you can inherit this with a shorthand for those rules,
```python
class MustPassUserRuleMixin(MustPassAuthCheckMixin):
    def check_rule(self):
             if not getattr(self,rule_name,None):
                 raise TypeError("must define rule_name on class"

              return test_rule(sefl.rule_name, self.request.user)
```

In the cases you want a view to check one rule for GET and a different one for POST, it might be easier to just code that logic in the view itself, again usually in the dispatch. Or you can write a mixin with 2 different check methods 🤷